mount/mount_idmapped_linux.go

```diff
@@ -87,80 +92,121 @@ func IDMapMount(source, target string, usernsFd int) (err error) {
 	return nil
 }
 
-// GetUsernsFD forks the current process and creates a user namespace using the specified
-// mappings.
-//
-// It returns:
-//  1. The file descriptor of the /proc/[pid]/ns/user of the newly
-//     created mapping.
-//  2. "Clean up" function that should be called once user namespace
-//     file descriptor is no longer needed.
-//  3. Usual error.
-func GetUsernsFD(uidmap, gidmap string) (_ int, _ func(), err error) {
-	var (
-		usernsFile       *os.File
-		pipeMap          [2]int
-		pid              uintptr
-		errno            syscall.Errno
-		uidMaps, gidMaps []syscall.SysProcIDMap
-	)
-
-	if uidMaps, err = parseIDMapping(uidmap); err != nil {
-		return -1, nil, err
-	}
-	if gidMaps, err = parseIDMapping(gidmap); err != nil {
-		return -1, nil, err
+// GetUsernsFD forks the current process and creates a user namespace using
+// the specified mappings.
+func GetUsernsFD(uidmap, gidmap string) (_usernsFD *os.File, _ error) {
+	uidMaps, err := parseIDMapping(uidmap)
+	if err != nil {
+		return nil, err
 	}
 
-	syscall.ForkLock.Lock()
-	if err = syscall.Pipe2(pipeMap[:], syscall.O_CLOEXEC); err != nil {
-		syscall.ForkLock.Unlock()
-		return -1, nil, err
+	gidMaps, err := parseIDMapping(gidmap)
+	if err != nil {
+		return nil, err
 	}
+	return getUsernsFD(uidMaps, gidMaps)
+}
+
+func getUsernsFD(uidMaps, gidMaps []syscall.SysProcIDMap) (_usernsFD *os.File, retErr error) {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
 
-	pid, errno = sys.ForkUserns(pipeMap)
-	syscall.ForkLock.Unlock()
+	pid, pidfd, errno := sys.ForkUserns()
 	if errno != 0 {
-		syscall.Close(pipeMap[0])
-		syscall.Close(pipeMap[1])
-		return -1, nil, errno
+		return nil, errno
 	}
 
-	syscall.Close(pipeMap[0])
+	pidFD := os.NewFile(pidfd, "pidfd")
```

Comment:
should probably check for pidFD == nil before continuing

Reply:
Hi @mikebrow ! Thanks for the comment.

The kernel returns the pidfd when we enable CLONE_PIDFD option. Basically, it won't be zero.
If the kernel doesn't support CLONE_PIDFD, the `ForkUserns` will return error.
So, I think we don't need to check pidFD == nil here.